src/hotspot/share/code/dependencyContext.cpp

```diff
@@ -68,16 +68,22 @@ int DependencyContext::mark_dependent_nmethods(DepChange& changes) {
   int found = 0;
   for (nmethodBucket* b = dependencies_not_unloading(); b != NULL; b = b->next_not_unloading()) {
     nmethod* nm = b->get_nmethod();
-    if (b->count() > 0 && nm->check_dependency_on(changes)) {
-      if (TraceDependencies) {
-        ResourceMark rm;
-        tty->print_cr("Marked for deoptimization");
-        changes.print();
-        nm->print();
-        nm->print_dependencies();
+    if (b->count() > 0) {
+      if (nm->is_marked_for_deoptimization()) {
+        // Also count already (concurrently) marked nmethods to make sure
+        // deoptimization is triggered before execution in this thread continues.
+        found++;
+      } else if (nm->check_dependency_on(changes)) {
+        if (TraceDependencies) {
+          ResourceMark rm;
+          tty->print_cr("Marked for deoptimization");
+          changes.print();
+          nm->print();
+          nm->print_dependencies();
+        }
+        changes.mark_for_deoptimization(nm);
+        found++;
       }
```

Comment:
A minor suggestion is to move the found++ from the individual branches to here, so there is only one.
Looks good!

Reply:
But that would increment `found` also in the cases where no method was marked for deoptimization, right?

Reply:
Yes, sorry, please ignore :)